pkg/lease/manager.go

```diff
@@ -0,0 +1,205 @@
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: pointer.Int32(int32(duration.Seconds())),
+			AcquireTime:          &microTimeNow,
+			RenewTime:            &microTimeNow,
+			LeaseTransitions:     pointer.Int32(0),
+		},
+	}
+
+	if err := l.Client.Create(ctx, lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+			nodeName := node.ObjectMeta.Name
+			key := apitypes.NamespacedName{Namespace: leaseNamespace, Name: nodeName}
+
+			if err := l.Client.Get(ctx, key, lease); err != nil {
+				return nil, false, err
+			}
+			return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false, nil
+}
+
+func (l *manager) updateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error) {
+	needUpdateLease := false
+	setAcquireAndLeaseTransitions := false
+	updateAlreadyOwnedLease := false
+
+	if lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity == holderIdentity {
+		needUpdateLease, setAcquireAndLeaseTransitions = needUpdateOwnedLease(lease, *currentTime, leaseDeadline)
+		if needUpdateLease {
+			updateAlreadyOwnedLease = true
+
+			log.Infof("renew lease owned by nmo setAcquireTime=%t", setAcquireAndLeaseTransitions)
+
+		}
+	} else {
+		// can't update the lease if it is currently valid.
+		if isValidLease(lease, currentTime.Time) {
+			return false, fmt.Errorf("can't update valid lease held by different owner")
+		}
+		needUpdateLease = true
+
+		log.Info("taking over foreign lease")
+		setAcquireAndLeaseTransitions = true
+	}
+
+	if needUpdateLease {
+		if setAcquireAndLeaseTransitions {
+			lease.Spec.AcquireTime = currentTime
+			if lease.Spec.LeaseTransitions != nil {
+				*lease.Spec.LeaseTransitions += int32(1)
+			} else {
+				lease.Spec.LeaseTransitions = pointer.Int32(1)
+			}
+		}
+		owner := makeExpectedOwnerOfLease(node)
+		lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
+		lease.Spec.HolderIdentity = &holderIdentity
+		lease.Spec.LeaseDurationSeconds = pointer.Int32(int32(leaseDuration.Seconds()))
+		lease.Spec.RenewTime = currentTime
+		if err := l.Client.Update(ctx, lease); err != nil {
+			log.Errorf("Failed to update the lease. node %s error: %v", node.Name, err)
+			return updateAlreadyOwnedLease, err
+		}
+	}
+
+	return false, nil
+}
+
+func (l *manager) invalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error {
+	log.Info("Lease object supported, invalidating lease")
+	nName := apitypes.NamespacedName{Namespace: leaseNamespace, Name: nodeName}
+	lease := &coordv1.Lease{}
+
+	if err := l.Client.Get(ctx, nName, lease); err != nil {
+
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	lease.Spec.AcquireTime = nil
+	lease.Spec.LeaseDurationSeconds = nil
+	lease.Spec.RenewTime = nil
+	lease.Spec.LeaseTransitions = nil
+
+	if err := l.Client.Update(ctx, lease); err != nil {
```

Comment:
I'm wondering why we don't delete the lease...?

Reply:
> this will eventually trigger the lease removal 

no one will delete the lease, if not we. There isn't something like a lease controller which deletes expired leases (to my best knowledge)

> However I'm not sure it'll be wise to make this kind of change of functionality in this PR (I see it as more of just exporting current lease functionality to common with a bit of a "face lift" and this sort of change feels more like an "open heart surgery" slightly_smiling_face )

Well, I see it otherwise. We are creating an API here, which will be used by multiple projects, and I don't think we want to update the API every few days or weeks when we notice that the API isn't ideal. I'd prefer to create a as-clean-as-possible API right away, and not a copy/paste of code from years ago.
We have the awsome chance to create the API while already having tests on a project which is going to use it. That's not a open heart surgery IMHO 🙂 
That's why I question the public methods very much, but also some impementation details. No matter who wrote the old code.

Reply:
>I'd prefer to create a as-clean-as-possible API right away, and not a copy/paste of code from years ago.

A full API examination/refactor wasn't my original intention, but since in any case we're almost there I'm happy to be onboard :sunglasses: 

- I think that the upside of updating the lease is saving resources in case of the lease is often used  (I would assume that constant delete / creation would take more resources than update) and a  faster acquirement if the lease.
- The downside is cluttering the etcd with unused resources (in case the lease is rarely used).

Let me know if you think otherwise, or have other considerations which I've missed and which approach do you prefer.

Reply:
IMHO the downside effect is bigger than the upside 🙂